Fix convention of PolynomialTensor basis change

aoowweenn opened the pull request:
In general, we rotate a matrix M to M' is to apply a rotation matrix R: M' = R @ M @ R.T
The corresponding Einstein notation is
M'^{p_1p_2} = R^{p_1}_{a_1} R^{p_2}_{b_1} M^{a_1a_2}

aoowweenn committed:
Fix convention of PolynomialTensor basis change

In general, we rotate a matrix M to M' is to apply a rotation matrix R:
M' = R @ M @ R.T
The corresponding Einstein notation is
M'^{p_1p_2} = R^{p_1}_{a_1} R^{p_2}_{b_1} M^{a_1a_2}

mpharrigan committed:
Merge branch 'master' into master

mhucka committed:
Merge branch 'main' into master

mhucka committed:
Run check/format-incremental --apply

mhucka commented:
@arettig Would you be able to take a look at this and check the technical correctness? (No rush.)

arettig committed:
Merge branch 'main' into pr805

arettig commented on src/openfermion/ops/representations/polynomial_tensor_test.py:
```diff
@@ -474,6 +474,40 @@ def test_rotate_basis_reverse(self):
         polynomial_tensor.rotate_basis(rotation_matrix_reverse)
         self.assertEqual(polynomial_tensor, want_polynomial_tensor)
 
+    def test_rotate_basis_90deg(self):
```
This test actually does not catch the bug you pointed out since for this specific case $$U^T M U$$ gives the same results as $$U M U^T$$. It would be a good idea to change this test to use a different basis rotation.

arettig replied:
Changed to a cyclical rotation test, which can distinguish between the incorrect previous code and the current fixed code.

arettig reviewed (changes requested):
This is a good catch. The `PolynomialTensor` rotation code currently does not match the expression in the [documentation](https://quantumai.google/openfermion/tutorials/circuits_1_basis_change). We should change this to match standard convention (and our own documentation), but this is a potentially breaking change for anyone using this function currently.

It would also be a good idea to change the 90 degree rotation test added in this PR to some different rotation as this test actually passes with the current (incorrect) implementation.

arettig committed:
Change 90 degree rotation test to cyclic rotation

The 90 degree rotation test passed even with previous erroneous code.
This new cyclical rotation (0->1, 1->2, 2->0) test will correctly test
the fixed bug.

arettig reviewed (approved):
LGTM
